### ISSUE TITLE
Specify program entrypoints during linking

### ISSUE THREAD

**Issue by PaulBone** [PlasmaLang/plasma#245]
The entrypoint is currently the function with the name "main". We'll remove this (but allow falling back to it in some circumstances). But we should allow the entrypoint to be specified during linking, so that any name could be used and the later-binding the more flexible (we don't need more one entrypoint per module). 

This requires that we can export functions from .p files first.

Things to test
 - [x] Single module with single entrypoint
 - [x] Multiple modules with single entrypoint chosen implicitly
 - [x] Single module with no entrypoint
 - [x] Single module with multiple entrypoints
 - [x] Multiple modules with multiple entrypoints
 - [x] What if the entrypoint module is not amoung those being linked.
 - [x] What if the entrypoint function is not found.
 - [x] Multiple programs generated by the same objects.
 - ~~Test that pure libraries (no entrypoints) also works.~~ (Will verify later)
 - [x] Invalid entrypoint (bad signature)
 - [x] When writing PZ assert that all entrypoints are also exported
 - [x] Copy candidates while linking
 - [x] Pretty-print entrypoints

**Comment by PaulBone** [comment]
Let's think about the rules for setting an entrypoint, first let's define two terms:

 * candidate:
   A function that is 1) exported 2) has a suitable signature.  The signature is either:

       func() uses IO -> Int

   or 

       func(args : List(String)) uses IO -> Int

   But we may include others in the future, eg should `env` be passed as a parameter also?

 * "main" candidate
   A candidate as above whose name is also "main".

   So the rules for setting an entrypoint should be:

       if linker command line specifies entrypoint:
          If found:
             Set the entrypoint
          else:
             Error!
       else if a entry module is specified on the command line:
          if found && has a "main" candidate
             Set the entrypoint
          else:
             Error!
       else if a single module is being compiled
          if it has a 'main" candidate.
             Set the entrypoint to this candidate
          else:
             Don't add an entrypoint (build a library).
       else:
          Don't add an entrypoint (build a library).

In the future we may add the concept that when compiling multiple modules, a subset of fhose modules are published as a library's interface.  If that set is only a single module, then check it for a "main" candidate.  If that set is zero and there is no specified entrypoint then that's an error.

**Comment by ACleverDisguise** [comment]
1. I would say that yes we should have a version that takes env as well.
2. Perhaps we might want to consider a compiler pragma to help the compiler make optimizations for candidates for main?
3. We should, with that pragma, permit multiple entry point candidates to be named in code, but only one is named at link time.  (Use case: making combined Windows GUI and command line apps with one standing in for "WinMain" and the other standing in for "main".)
4. In the case of some libraries, there's a "DLLMain"-like function that may come into play.  Do we want some way to name that?
5. If so, perhaps we need to be explicit on the command line if we're building an executable or a library with the entry point, if named in either case (going to defaults if not), being part of the switch like:

… --build-executable[=entry_point]

**Comment by PaulBone** [comment]
>     1. I would say that yes we should have a version that takes env as well.

I like this because it feels weird that command line arguments are passed in when env is not. But they're both part of "starting a program" (whether that's fork/exec/etc.

>     2. Perhaps we might want to consider a compiler pragma to help the compiler make optimizations for candidates for main?
>
>     3. We should, with that pragma, permit multiple entry point candidates to be named in code, but only one is named at link time.  (Use case: making combined Windows GUI and command line apps with one standing in for "WinMain" and the other standing in for "main".)

As it is now we export a function with the `export` keyword.  How about `entrypoint` which has the same meaning as export plus marks the function as a candidate (so the linker may choose among those marked as `entrypoint`.

This is at the risk of adding more keywords to the language and more things for a novice to think of.  However that may be more explicit than the special name "main" being the default. We can default to "If there is exactly one entrypoint then just use it".

And yes, I'm thinking of `main`/`WinMain` as an example of when you need different entrypoints.  I didn't know about `DllMain`.  But also if Plasma is used in different environments.  Eg compiled to Wasm in the browser, it may have different entrypoints (the browser is a bad example since it's more like a library there, responding to events).  Different environments may wish to call programs with different entrypoint signatures.
 
>     4. In the case of some libraries, there's a "DLLMain"-like function that may come into play.  Do we want some way to name that?

I don't see why not. My gut reaction said that this was more like handling an event "we got loaded". I think that's because AIUI this call will exit again without terminating the program, and now the library's exported functions may be called.  I think that Linux/SysV/ELF (not sure which specifies it) has some kind of "on load" and "on unload".  The latter is important for freeing resources when the library is loaded at runtime (not execution time).  We found that out the hard way since libpam loads and unloads it's authentication modules for each authentication!

Anyway, it feels more like an event since it exits without terminating the program.  However for that to make sense we'd need to register it to handle the event, why not just make entrypoints able to work like this? there's no rule against it.  This means we may have multiple entrypoints per program/library to handle load/unload of dynamic libraries, in addition to your multiple entrypoints for `main` vs `WinMain` which I only briefly thought of before.  Maybe this is the kind of thing that we should avoid engineering now but just not paint ourselves into a corner.

>     5. If so, perhaps we need to be explicit on the command line if we're building an executable or a library with the entry point, if named in either case (going to defaults if not), being part of the switch like:
> 
> 
> … --build-executable[=entry_point]
> … --build-static-library
> … --build-dynamic-library[=entry_point]
> 
> And in the case of dynamic libraries there are bizarre hybrid executables that are Windows CLI, Windows GUI, and Windows DLL files all in the same binary. Is this something we want to consider?

Maybe.  Keeping in mind that users won't normally use this interface.  They'll write some kind of project file (#253) that will specify the name of their program/library, its entrypoints, any modules to export (for libraries), and any dependencies & their versions.

    # Project file for my program.  (`hello.plz`)
    Name: Hello
    Type: program
    Depends-on: left-pad
    Modules: Hello

I almost wrote in Type that it could be a native program or bytecode program.  But whether we do a bytecode/native/optimised/profiling/etc build should be part of the command line (or something) and not part of saying what makes up the project.

**Comment by PaulBone** [comment]
>     3\. We should, with that pragma, permit multiple entry point candidates to be named in code, but only one is named at link time.  (Use case: making combined Windows GUI and command line apps with one standing in for "WinMain" and the other standing in for "main".)

I forgot to say.  My previous idea for this was that a programmer would put these different entrypoints in different modules (containing only that entrypoint and any code unique to that entrypoint). Then at link time they choose which module (and therefore entrypoint) to use.  Your proposal is a bit more flexible and could be useful when something actually needs two entrypoints,  like a DllMain called when the library is loaded, and another one called when it exists.

**Comment by PaulBone** [comment]
>     2\. Perhaps we might want to consider a compiler pragma to help the compiler make optimizations for candidates for main?

Also how do you mean "optimisations"?  Like "don't optimise away exported code" because we won't do that if it's exported.

**Comment by ACleverDisguise** [comment]
> Also how do you mean "optimisations"? Like "don't optimise away exported code" because we won't do that if it's exported.

Optimizing away exported code if it's not used in a final assembly is exactly what I was thinking.  Why wouldn't you optimize away an exported function that's not used?

But other optimizations may include things like the way stack frames are set up, etc.

**Comment by PaulBone** [comment]
> > Also how do you mean "optimisations"? Like "don't optimise away exported code" because we won't do that if it's exported.
> 
> Optimizing away exported code if it's not used in a final assembly is exactly what I was thinking. Why wouldn't you optimize away an exported function that's not used?
> 
> But other optimizations may include things like the way stack frames are set up, etc.

Yep, just clarifying.

There's already the concept of exported.  If something is exported then it can't be optimised away, entrypoints are no different.  What we might want to do, mostly in the future:

 * If a function is exported, but that module itself isn't exported than that function can be optimised away if it has no references (and other things that now have no references since removing that function).
 * We could make it possible for something to be an entrypoint without being exported.  That is other modules can't call it, but the program still starts there.
 * Probably want to make exported and "exported to foreign code" separate concepts too, like you said, how stack frames are setup.

Generally "all the things Plasma could do" has a huge scope.  I'm trying not to think about it that much including optimisation.  and just get features in - while not painting ourselves into a corner.
